cogs/utils/arg_converter.py

```diff
@@ -0,0 +1,138 @@
+            if argument[0] != a:
+                raise commands.BadArgument("Argument doesn't start with 'a'")
+            result = argument[1:]
+            return result
+
+    A converter could also simply check that an argument is of a certain type
+    or has certain attributes and return the argument itself with no
+    conversion, which would act like a pseudo-typecheck. It could also
+    call another converter, for example stripping a certain argument and
+    and trying to convert the remaining string to a discord member.
+
+    Example use:
+
+    from discord.ext.commands import MemberConverter, TextChannelConverter
+    class StartsByAConverter(commands.Converter):
+        (as above)
+    @commands.command()
+    async def example(self, ctx, *args):
+      arg_converter = ArgConverter({"member": (MemberConverter(), None),
+                                    "channel": (TextChannelConverter(), )
+                                    "string_A": (FromConverter(), "hi"))
+    try:
+        args_dict = await arg_converter.convert(ctx, args)
+    except commands.BadArgument as err:
+        await ctx.send(err)
+        return
+    await ctx.send(f"{str(args_dict["member"])}, "
+                   f"{args_dict["channel"]}, "
+                   f"{args_dict["string_A"]}")
+
+    Here are example inputs and what is sent in the server:
+    `?example #channel`: None, #channel, hi
+    `?example aword @someone #channel`: someone#1234, #channel, word
+    `?example aword @someone"`:
+    Invalid input: Missing required argument "channel"
+    `?example #channel aword @someone woo`:
+    Invalid input: Argument "woo" could not be converted
+    """
+    def __init__(self, converters_dict):
+        self.converters_dict = converters_dict
+
+    async def convert(self, ctx, arguments):
+        """
+        Match the arguments to the converters. If a converter is used,
+        doesn't try to match it again. All required arguments must be
+        given, and all arguments must be able to be converted.
+
+        Order doesn't matter, except if there are multiple converters that
+        match with the same argument, in which case the first one in the
+        dictionary is used. This could lead to unexpected results:
+        for example, it doesn't matter if this is to catch two discord users,
+        but if there was a converter meant to catch any string, then a
+        converter to catch an uppercase string, then the arguments
+        ["HI", "hello"] will raise an exception, since the "any" string
+        converter is the one used with "HI", and then the only remaining
+        converter is the uppercase one, and so "hello" is not converted.
+
+        Returns a dictionary with the variables as keys, and their values.
+        """
+        remaining_vars = self.converters_dict.copy()
+        # initialize the converted dict with None values
+        converted_arguments_dict = {}
+
+        for arg in arguments:
```

Comment:
how does this deal with string ordering and stuff? e.g. ordered string arguments

Reply:
The first converter in the dict is used as written in comment higher up. It is now clearer that the dict should be ordered in 8104c27